drivers/sensor: fix some ST drivers dependency to HAL_ST module

avisconti opened the pull request:
In order to avoid build failures as described in issue #67242, make ST drivers using HAL_ST module dependent to HAL_STMEMSC and HAS_STLIB, which needs to be configured in all samples referring to them.

Fixes #67242

nordicjm reviewed (changes requested):
This is not correct, zephyr has module support in cmake, so your hal_st module should have this set and you should be able to know in Kconfig (as Kconfig runs after cmake and has the zephyr modules set) if a zephyr module is set, and I'm not seeing any of that used here. A board Kconfig fragment cannot know if a module is present

erwango commented on samples/boards/96b_argonkey/sensors/prj.conf:
```diff
@@ -5,5 +5,6 @@ CONFIG_I2C=y
 CONFIG_LED=y
 CONFIG_GPIO=y
 CONFIG_SENSOR=y
+CONFIG_HAS_STMEMSC=y
```
Can it somehow be done directly in hal_st module ?
So that the dependency is resolved directly and simply by having the module in the manifest ?

avisconti replied:
I moved back in the driver itself, where also the ```depends on ZEPHYR_ST_MODULE``` lies

nordicjm commented:
`ZEPHYR_ST_MODULE` is what you use to know if the module is present or not

nordicjm commented on drivers/sensor/hts221/Kconfig:
```diff
@@ -5,9 +5,9 @@ menuconfig HTS221
-	select HAS_STMEMSC
 	select USE_STDC_HTS221
 	help
 	  Enable driver for HTS221 I2C/SPI-based temperature and humidity sensor.
```
Assuming hts221 needs the hal_st repo present, here is a rough idea:
```
menuconfig HTS221
	bool "HTS221 temperature and humidity sensor"
	default y
	depends on DT_HAS_ST_HTS221_ENABLED
	depends on ZEPHYR_ST_MODULE
	select I2C if $(dt_compat_on_bus,$(DT_COMPAT_ST_HTS221),i2c)
	select SPI if $(dt_compat_on_bus,$(DT_COMPAT_ST_HTS221),spi)
	select USE_STDC_HTS221
	help
	  Enable driver for HTS221 I2C/SPI-based temperature and humidity sensor.
```

avisconti replied:
Ok, need to refer to ```ZEPHYR_ST_MODULE```. Thx

avisconti committed:
drivers/sensor: fix some ST drivers dependency to HAL_ST module

In order to avoid build failures as described in issue #67242,
make all ST drivers using HAL_ST module dependent to HAL_STMEMSC
and HAS_STLIB libs, which need to be configured in all samples
referring to them.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

nordicjm commented on modules/Kconfig.st:
```diff
@@ -3,10 +3,10 @@
 config HAS_STLIB
-	bool
+	bool "ST Micro time-of-flight library"
 
 config HAS_STMEMSC
-	bool
+	bool "ST Micro STdC (STMEMSC) library"
```
This should not be selectable (revert)

avisconti replied:
But I need to enable HAS_STMEMSC. I saw for example that you do ```select HAS_NORDIC_DRIVERS``` even if seems not selectable. I'm a bit confused here.

nordicjm reviewed (commented):
Seems a module needs to be added to the `modules` folder, would need `  kconfig-ext: True` in hal_st repo and then `modules/st/Kconfig` would have something like:
```
config ZEPHYR_ST_MODULE
        bool
```

avisconti commented:
> Seems a module needs to be added to the `modules` folder, would need ` kconfig-ext: True` in hal_st repo and then `modules/st/Kconfig` would have something like:
> 
> ```
> config ZEPHYR_ST_MODULE
>         bool
> ```

Mmmh, I'm quite new here. What about ```cmake-ext: True``` purpose? Do I need it as well?

nordicjm commented:
> > Seems a module needs to be added to the `modules` folder, would need ` kconfig-ext: True` in hal_st repo and then `modules/st/Kconfig` would have something like:
> > ```
> > config ZEPHYR_ST_MODULE
> >         bool
> > ```
> 
> Mmmh, I'm quite new here. What about `cmake-ext: True` purpose? Do I need it as well?

No, that is only needed if you had the CMakeLists.txt file for hal_st inside of the zephyr tree

avisconti commented:
> > > Seems a module needs to be added to the `modules` folder, would need ` kconfig-ext: True` in hal_st repo and then `modules/st/Kconfig` would have something like:
> > > ```
> > > config ZEPHYR_ST_MODULE
> > >         bool
> > > ```
> > 
> > 
> > Mmmh, I'm quite new here. What about `cmake-ext: True` purpose? Do I need it as well?
> 
> No, that is only needed if you had the CMakeLists.txt file for hal_st inside of the zephyr tree

ok, done. Thx

nordicjm reviewed (dismissed):
Brilliant, thanks!

fabiobaltieri commented on west.yml:
```diff
@@ -224,7 +224,7 @@ manifest:
         - hal
     - name: hal_st
-      revision: fb8e79d1a261fd02aadff7c142729f1954163cf3
+      revision: pull/18/head
```
`0643d20ae85b32c658ad11036f7c964a860ddefe`

avisconti replied:
done

avisconti commented on modules/Kconfig:
```diff
@@ -31,7 +31,6 @@ source "modules/Kconfig.nxp_s32"
 source "modules/Kconfig.simplelink"
 source "modules/Kconfig.sof"
-source "modules/Kconfig.st"
 source "modules/Kconfig.stm32"
```
> Genuine question: No need to add `modules/hal_st/Kconfig` here ?

had the same question myself, but it seems that others (like hal_nxp) don't have it.

fabiobaltieri replied:
looks like the build systems adds that for you in `build/Kconfig/Kconfig.modules`

fabiobaltieri replied:
...not sure how it gets included when the module is not checked out though, but it seems like it does.

fabiobaltieri replied:
oh, this, https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/Kconfig#L118-L123, yeah so if the hal_st module is there it gets included from the generated file and the symbols are usable, otherwise it gets included by the wildcard with all symbols excluded, which is exactly the same because all symbols are just placeholders but I guess it's the right thing to do anyway

avisconti replied:
> oh, this, https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/Kconfig#L118-L123, yeah so if the hal_st module is there it gets included from the generated file and the symbols are usable, otherwise it gets included by the wildcard with all symbols excluded, which is exactly the same because all symbols are just placeholders but I guess it's the right thing to do anyway

Thanks for explanation!
